[CanvasNewtSWT] call Widget.dispose() to release SWT resources and clean up in SWT-project

HannesWell opened the pull request:
If a Chart is disposed and as a follow up `CanvasSWT.dispose()` was called the canvas remained in its parent as a white or black canvas because `super.dispose()` was not called so the SWT-disposal process was not initiated.

Amendment:
Additionally I did some code formatting in the SWT-project since the code was formatted inconsistently itself and to the rest of jzy3d and some clean up and avoids of code duplications.
Furthermore I use the JFace ResourceManager in `CanvasSWT` to handle `Color` and `Font` resources in order to solved the `TODO` annotation, complaining about the SWT-resource handling.

As a result of the discussion I moved the jzy3D resource release into the disposeListener of `CanvasSWT` and removed dispose. Now if dispose is called for a `CanvasSWT` instance `Widget.dispose()` is called directly and within the disposal process the jzy3d-resources are released by the dispose-listener. So the initial issue of this PR is resolved with that and release of jzy3d-resources is still ensured.

jzy3d commented:
I would like @jacobfilik to give its opinion on that change, as he is the person who wrote this class!

HannesWell commented:
Yes, sure!

To test the dispose behavior I modified the SWTDemo class and provide two buttons so the user can add and remove charts. Without my change the Composite of the canvas remains in UI as a white rectangle (under Win10). 
Additionally the layout behaves wrong. Actually the canvases should be placed side by side. And because the layout assumes that the swt.Control of the canvas is gone and reduces its number of columns, so that the number of columns is equal to the number of (active) canvases. But since the canvas remains a second row is created. 

With the submitted change everything works perfectly fine.

The following is the modified code of the SWT-Demo I used:

> package org.jzy3d.analysis;
> 
> import java.util.ArrayList;
> import java.util.List;
> 
> import org.eclipse.swt.SWT;
> import org.eclipse.swt.layout.FillLayout;
> import org.eclipse.swt.layout.GridData;
> import org.eclipse.swt.layout.GridLayout;
> import org.eclipse.swt.widgets.Button;
> import org.eclipse.swt.widgets.Composite;
> import org.eclipse.swt.widgets.Control;
> import org.eclipse.swt.widgets.Display;
> import org.eclipse.swt.widgets.Shell;
> import org.jzy3d.chart.Chart;
> import org.jzy3d.chart.ChartLauncher;
> import org.jzy3d.chart.Settings;
> import org.jzy3d.chart.swt.SWTChartComponentFactory;
> import org.jzy3d.colors.Color;
> import org.jzy3d.colors.ColorMapper;
> import org.jzy3d.colors.colormaps.ColorMapRainbow;
> import org.jzy3d.maths.Range;
> import org.jzy3d.plot3d.builder.Builder;
> import org.jzy3d.plot3d.builder.Mapper;
> import org.jzy3d.plot3d.builder.concrete.OrthonormalGrid;
> import org.jzy3d.plot3d.primitives.Shape;
> 
> public class SWTDemo {
> 
>     public static void main(String[] args) {
>         Shape surface = getSurface();
> 
>         // Create a chart
> 
>         Settings.getInstance().setHardwareAccelerated(true);
> 
>         Display display = new Display();
>         Shell shell = new Shell(display);
>         shell.setLayout(new FillLayout());
> 
>         Composite container = new Composite(shell, SWT.NONE);
>         container.setLayout(new GridLayout(2, false));
> 
>         Button addButton = new org.eclipse.swt.widgets.Button(container, SWT.PUSH);
>         addButton.setLayoutData(new GridData(SWT.BEGINNING, SWT.CENTER, false, false));
>         addButton.setText("Add chart");
> 
> 
>         Button removeButton = new Button(container, SWT.PUSH);
>         removeButton.setText("Remove first chart");
>         removeButton.setLayoutData(new GridData(SWT.BEGINNING, SWT.CENTER, false, false));
> 
>         Composite chartContainer = new Composite(container, SWT.BORDER);
>         chartContainer.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true, 2, 1));
>         GridLayout layout = new GridLayout(0, true);
>         chartContainer.setLayout(layout);
> 
>         List<Chart> charts = new ArrayList<>();
> 
>         addButton.addListener(SWT.Selection, e -> {
>             charts.add(addChart(chartContainer, getSurface()));
>             layout(chartContainer);
>         });
> 
>         removeButton.addListener(SWT.Selection, e -> {
>             if (!charts.isEmpty()) {
>                 charts.remove(0).dispose();
>                 ((GridLayout) chartContainer.getLayout()).numColumns--;
>                 layout(chartContainer);
>             }
>         });
> 
>         Chart chart = addChart(chartContainer, surface);
>         charts.add(chart);
> 
>         runShell(display, shell);
>     }
> 
>     private static void layout(Composite chartContainer) {
>         for (Control child : chartContainer.getChildren()) {
>             requestLayout(child);
>         }
>     }
> 
>     private static Chart addChart(Composite chartContainer, Shape surface) {
>         Chart chart = SWTChartComponentFactory.chart(chartContainer);
>         chart.getScene().getGraph().add(surface);
> 
>         ChartLauncher.openChart(chart); // add mouse controllers
> 
>         ((GridLayout) chartContainer.getLayout()).numColumns++;
>         Control[] children = chartContainer.getChildren();
>         children[children.length - 1].setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));
> 
>         requestLayout(chartContainer);
>         return chart;
>     }
> 
> 
>     private static void requestLayout(Control c) {
>         c.getShell().layout(new Control[] { c }, SWT.DEFER);
>     }
> 
>     private static Shape getSurface() {
>         Mapper mapper = new Mapper() {
>             @Override
>             public double f(double x, double y) {
>                 return x * Math.sin(x * y);
>             }
>         };
> 
>         // Define range and precision for the function to plot
>         Range range = new Range(-3, 3);
>         int steps = 80;
> 
>         // Create the object to represent the function over the given range.
>         final Shape surface = Builder.buildOrthonormal(new OrthonormalGrid(range, steps, range, steps), mapper);
>         surface.setColorMapper(
>                 new ColorMapper(new ColorMapRainbow(), surface.getBounds().getZmin(), surface.getBounds().getZmax(), new Color(1, 1, 1, .5f)));
>         surface.setFaceDisplayed(true);
>         surface.setWireframeDisplayed(false);
>         return surface;
>     }
> 
>     private static void runShell(Display display, Shell shell) {
>         shell.setText("SWT chart Demo");
>         shell.setSize(800, 600);
>         shell.open();
> 
>         while (!shell.isDisposed()) {
>             if (!display.readAndDispatch()) {
>                 display.sleep();
>             }
>         }
>         display.dispose();
>     }
> }

jacobfilik commented:
Hi, yes should probably call dispose on the super to put the widget in a disposed state so this is an improvement.

Outside of this change I think the resources should be cleaned up in a dispose listener rather than in the dispose method, which I should probably look at testing/fixing, but its not critical, I think it only catches some edge cases better.

HannesWell commented:
Yes you are right.
I thought `dispose()` is called every time a `Widget` is disposed either directly or because a parent `Widget` is disposed. But I was wrong. If a parent `Widget` is disposed `dispose()` of the children is not called. So doing the resource release in a dispose-listener ensures that the resources are released every time the widget is disposed. 
I think the best solution is to move the release of jzy3d-resources into a dispose-listener and remove the dispose override from CanvasNewtSWT.

HannesWell commented:
Additionally I added some changes that I originally wanted to submit in a follow-up PR but since we are discussing here already I submit them now.

These commits include formatting of the code in the SWT-project since the code was formatted inconsistently itself and to the rest of jzy3d and some clean up and avoids of code duplications.
Furthermore I use the JFace ResourceManager in `CanvasSWT` to handle `Color` and `Font` resources in order to solved the `TODO` annotation, complaining about the SWT-resource handling.

And as mentioned before I moved the jzy3D resource release into the disposeListener of `CanvasSWT` and removed dispose. Now if dispose is called for a `CanvasSWT` instance `Widget.dispose()` is called directly and within the disposal process the jzy3d-resources are released by the dispose-listener. So the initial issue of this PR is resolved with that and release of jzy3d-resources is still ensured.
